src/index.js

```diff
@@ -3,19 +3,32 @@ import queryString from 'query-string';
 const defaultUrlRegExp = /^(\w+:\/\/[^/?]+)?(.*?)(\?.+)?$/;
 const protocolRelativeUrlRegExp = /^(\/\/[^/?]+)(.*?)(\?.+)?$/;
 
-function splitUrl(partsStr, { protocolRelative }) {
+const pipe = (...fns) => fns.reduce((f, g) => (...args) => g(f(...args)));
+
+function splitUrl(partsStr, { protocolRelative, trailingSlash, leadingSlash }) {
     const match = (protocolRelative && partsStr.match(protocolRelativeUrlRegExp)) ||
                    partsStr.match(defaultUrlRegExp) ||
                   [];
 
     const beforePathname = match[1] || '';
-    const pathname = (match[2] || '')
-    // Remove leading slashes
-    .replace(/^\/+/, '')
-    // Remove trailing slashes
-    .replace(/\/+$/, '')
-    // Normalize consecutive slashes to just one
-    .replace(/\/+/g, '/');
+
+    const shouldKeep = (opt) => opt === 'keep';
+
+    const removeLeadingSlashes = (leadingSlash) =>
+        (s) => shouldKeep(leadingSlash) ? s : s.replace(/^\/+/, '');
```

Comment:
I would name `s` as `pathname` instead to be more clear.

Reply:
This is not needed anymore, because is only taking the `leadingSlash` now